### PR TITLE
feature: Add Inicial Address API structure

### PR DESCRIPTION
**Pull request by gdseabra**
Created a REST API for the Address Entity that allow the following operations:
- Fetch all client's addresses
- Fetch a specific address by id
- Insert a client's address
- Alter a specific address in db
- Remove an address

### CODE REVIEW

File: src/main/java/com/matera/cadastrocentral/address/Address.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
+package com.matera.cadastrocentral.address;
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.Id;
+import java.util.UUID;
+
+@Entity
+public class Address {
+
+    private @Id @GeneratedValue UUID address_id;
```
Comment (rgueltamatera) [inline src/main/java/com/matera/cadastrocentral/address/Address.java]: can you please remove the @GeneratedValue  anotation? when we use UUIDs for entity identification we create it on application not on database

File: src/main/java/com/matera/cadastrocentral/address/Address.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    private String state;
+    private String zip_code;
+    private String complement;
+    private String reference;
+    private String city;
+
+
+    public Address() {}
+
+    public Address(UUID clientId, String street_name, String district, String state, String zip_code, String complement, String reference) {
```
Comment (rgueltamatera) [inline src/main/java/com/matera/cadastrocentral/address/Address.java]: I don't think this constructor is used anywere, can we get rid of it ?

File: src/main/java/com/matera/cadastrocentral/address/Address.java (hunk 1 above)
Comment (gdseabra) [reply src/main/java/com/matera/cadastrocentral/address/Address.java]: Sorry, I thought I had to let the database generate it automatically. Should I add a UUID uuid = UUID.randomUUID() in the constructor?

File: src/main/java/com/matera/cadastrocentral/address/AddressController.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import com.sun.org.apache.bcel.internal.generic.NEW;
+import org.apache.catalina.filters.AddDefaultCharsetFilter;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+import java.util.UUID;
+
+@RestController
+@RequestMapping("/addresses")
```
Comment (rgueltamatera) [inline src/main/java/com/matera/cadastrocentral/address/AddressController.java:12]: Great job @gdseabra, I think it's good as it is for now, i will create a refactor ticket so we can call the API by "/clients/{clientId}/addresses..." 
I forgot to deescribe it should be client specific, so the blame is on me :facepalm:

File: src/main/java/com/matera/cadastrocentral/address/AddressService.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    public void deleteAddress(UUID id) {
+        addressRepository.deleteById(id);
+    }
+
+    public Address alterAddress(UUID id, Address newAddress) {
+        Address alteredAddress = addressRepository.findById(id).orElse(null);
+
+        alteredAddress.setReference(Optional.ofNullable(newAddress.getReference()).orElse(alteredAddress.getReference()));
```
Comment (rgueltamatera) [inline src/main/java/com/matera/cadastrocentral/address/AddressService.java:39]: if alteredAddress is null we will get a NPE(NullPointerException) here, maybe check if its null then return null before reaching this code is a good idea

File: src/main/java/com/matera/cadastrocentral/address/Address.java (hunk 1 above)
Comment (rgueltamatera) [reply src/main/java/com/matera/cadastrocentral/address/Address.java]: you can create it in the service or you can add a PrePersist annotaded method on the entity, like:

```
@PrePersist
public void prePersist() {
    this.addressId = UUID.randomUUID();
}
```

File: src/main/java/com/matera/cadastrocentral/address/Address.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+import javax.persistence.GeneratedValue;
+import javax.persistence.Id;
+import java.util.UUID;
+
+@Entity
+public class Address {
+
+    private @Id @GeneratedValue UUID address_id;
+    private UUID clientId;
+    private String street_name;
```
Comment (rgueltamatera) [inline src/main/java/com/matera/cadastrocentral/address/Address.java]: can you change the attributes to use camelCase ?

File: src/main/java/com/matera/cadastrocentral/address/Address.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+    public String getCity() {
+        return city;
+    }
+
+    public void setCity(String city) {
+        this.city = city;
+    }
+
+    public String getStreet_name() {
```
Comment (rgueltamatera) [inline src/main/java/com/matera/cadastrocentral/address/Address.java]: Also rename methods to use camelCase

File: src/main/java/com/matera/cadastrocentral/address/AddressController.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+    @Autowired
+    private AddressService addressService;
+
+    @PostMapping
+    Address newAddress(@RequestBody Address newAddress) {
+        return addressService.insertAddress(newAddress);
+    }
+
+    @GetMapping("/{address_id}")
```
Comment (rgueltamatera) [inline src/main/java/com/matera/cadastrocentral/address/AddressController.java]: uou can use camelCase here too, like: {addressId}

**Comment by gdseabra** [comment]
Made @rgueltamatera's  suggested changes:
- Switch to camelCase
- Remove @GeneratedValue and add @PrePersist
- Remove useless constructor
- Check alteredAddress  for null

File: src/main/java/com/matera/cadastrocentral/address/Address.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.Id;
+import javax.persistence.PrePersist;
+import java.util.UUID;
+
+@Entity
+public class Address {
+
+    private @Id @GeneratedValue UUID addressId;
```
Comment (izumizawa) [inline src/main/java/com/matera/cadastrocentral/address/Address.java]: Robin mentioned earlier that @GeneratedValue isn't necessary here

**Review by izumizawa — APPROVED** [review]

File: src/main/java/com/matera/cadastrocentral/address/Address.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    private String reference;
+    private String city;
+
+
+    @PrePersist
+    public void prePersist() {
+        this.addressId = UUID.randomUUID();
+    }
+
+    public Address(UUID clientId, String streetName, String district, String state, String zipCode, String complement, String reference) {
```
Comment (rgueltamatera) [inline src/main/java/com/matera/cadastrocentral/address/Address.java]: looks like the constructor is not used anywere, can you remove it?

Comment (gdseabra) [reply src/main/java/com/matera/cadastrocentral/address/Address.java]: How will the application know how to create the instance if I delete the constructor? or does Spring automaticaly figure it out?

Comment (rgueltamatera) [reply src/main/java/com/matera/cadastrocentral/address/Address.java]: spring figure it out :)

**Review by rgueltamatera — APPROVED** [review]

**Pull request merged**